extensions/sql-database-projects/src/typings/vscode-mssql.d.ts

```diff
@@ -0,0 +1,291 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+declare module 'vscode-mssql' {
+    /**
+     * Covers defining what the vscode-mssql extension exports to other extensions
+     *
+     * IMPORTANT: THIS IS NOT A HARD DEFINITION unlike vscode; therefore no enums or classes should be defined here
```

Comment:
For my own learning: what's a "hard definition" (as opposed to a soft one)?

Reply:
Hard definition is something that is an actual javascript object. In the case of typings files those are only runtime constructs and so if you have anything that needs to exist at runtime (such as a class definition or non-const enum) then it can't live here.

azdata.d.ts/vscode.d.ts are "special" in that at runtime an object is injected into the process that implements the module being defined - hence why we can have classes and whatnot defined there.